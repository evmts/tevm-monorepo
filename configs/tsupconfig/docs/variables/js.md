**@tevm/tsupconfig** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > js

# Variable: js

> **`const`** **js**: `Options`

## Source

[js.js:3](https://github.com/evmts/tevm-monorepo/blob/main/configs/tsupconfig/src/js.js#L3)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
