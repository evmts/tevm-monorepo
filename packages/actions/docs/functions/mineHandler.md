[**@tevm/actions**](../README.md) • **Docs**

***

[@tevm/actions](../globals.md) / mineHandler

# Function: mineHandler()

> **mineHandler**(`client`, `options`?): [`MineHandler`](../type-aliases/MineHandler.md)

## Parameters

• **client**: `TevmNode`\<`"fork"` \| `"normal"`, `object`\>

• **options?** = `{}`

• **options.throwOnFail?**: `undefined` \| `boolean`

whether to default to throwing or not when errors occur

## Returns

[`MineHandler`](../type-aliases/MineHandler.md)

## Defined in

[packages/actions/src/Mine/mineHandler.js:16](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/Mine/mineHandler.js#L16)
