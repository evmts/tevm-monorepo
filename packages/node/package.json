{
	"name": "@tevm/node",
	"version": "2.0.0-next.108",
	"private": false,
	"description": "Client implementations of Tevm",
	"keywords": [
		"solidity",
		"sol",
		"typescript",
		"web3",
		"blockchain"
	],
	"repository": {
		"type": "git",
		"url": "https://github.com/evmts/tevm-monorepo.git",
		"directory": "core/node"
	},
	"license": "MIT",
	"contributors": [
		"Will Cory <willcory10@gmail.com>"
	],
	"type": "module",
	"exports": {
		"./package.json": "./package.json",
		".": {
			"import": {
				"types": "./dist/index.d.ts",
				"default": "./dist/index.js"
			},
			"require": {
				"types": "./dist/index.d.cts",
				"default": "./dist/index.cjs"
			}
		}
	},
	"main": "dist/index.cjs",
	"module": "dist/index.js",
	"types": "dist/index.d.ts",
	"files": [
		"dist",
		"src",
		"!src/**/*.spec.ts",
		"types"
	],
	"scripts": {
		"all": "pnpm i  && bun run build && bun lint && bun format && bun test:run && bun generate:docs && bun lint:deps && bun lint:package",
		"build": "nx run-many --targets=build:dist,build:types --projects=@tevm/node",
		"build:dist": "tsup",
		"build:types": "tsup --dts-only",
		"clean": "rm -rf node_modules && rm -rf artifacts && rm -rf dist && rm -rf cache",
		"format": "biome format . --write",
		"format:check": "biome format .",
		"generate:docs": "typedoc",
		"lint": "biome check . --write --unsafe",
		"lint:check": "biome check . --verbose",
		"lint:deps": "bunx depcheck",
		"lint:package": "bunx publint --strict && attw --pack",
		"package:up": "pnpm up --latest",
		"test": "vitest --watch",
		"test:coverage": "vitest --coverage",
		"test:run": "vitest",
		"typecheck": "tsc --noEmit"
	},
	"dependencies": {
		"@tevm/address": "workspace:^",
		"@tevm/block": "workspace:^",
		"@tevm/blockchain": "workspace:^",
		"@tevm/common": "workspace:^",
		"@tevm/evm": "workspace:^",
		"@tevm/jsonrpc": "workspace:^",
		"@tevm/logger": "workspace:^",
		"@tevm/predeploys": "workspace:^",
		"@tevm/receipt-manager": "workspace:^",
		"@tevm/state": "workspace:^",
		"@tevm/sync-storage-persister": "workspace:^",
		"@tevm/tx": "workspace:^",
		"@tevm/txpool": "workspace:^",
		"@tevm/utils": "workspace:^",
		"@tevm/vm": "workspace:^"
	},
	"devDependencies": {
		"@tevm/test-utils": "workspace:^",
		"@tevm/tsconfig": "workspace:^",
		"@tevm/tsupconfig": "workspace:^"
	},
	"peerDependencies": {
		"viem": "^2.14.2"
	},
	"publishConfig": {
		"access": "public"
	},
	"sideEffect": false
}
