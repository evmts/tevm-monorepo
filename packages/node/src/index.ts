export { createTevmNode } from './createTevmNode.js'
export type { TevmNode } from './TevmNode.js'
export type { TevmNodeOptions } from './TevmNodeOptions.js'
export type { BaseClientOptions } from './BaseClientOptions.js'
export type { CustomPrecompile } from './CustomPrecompile.js'
export type { Extension } from './Extension.js'
export type { Hardfork } from './Hardfork.js'
export type {
	MiningConfig,
	AutoMining,
	ManualMining,
	IntervalMining,
} from './MiningConfig.js'
export { prefundedAccounts, GENESIS_STATE } from './GENESIS_STATE.js'
export type { Filter, FilterType } from './Filter.js'
export {
	ProviderRpcError,
	type EIP1193Events,
	type EIP1193EventMap,
	type ProviderMessage,
	type EIP1193EventEmitter,
	type ProviderConnectInfo,
} from './EIP1193EventEmitterTypes.js'
