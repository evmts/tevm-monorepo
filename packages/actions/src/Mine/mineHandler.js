import { InternalError, MisconfiguredClientError, UnreachableCodeError } from '@tevm/errors'
import { bytesToHex, hexToBytes } from '@tevm/utils'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { emitEvents } from './emitEvents.js'
import { validateMineParams } from './validateMineParams.js'

// TODO Errors can leave us in bad states

/**
 * @param {import("@tevm/node").TevmNode} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('./MineHandlerType.js').MineHandler}
 */
export const mineHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail ?? true, ...params } = {}) => {
		client.logger.debug({ throwOnFail, ...params }, 'mineHandler called with params')
		const errors = validateMineParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, { errors })
		}
		const { interval = 1, blockCount = 1 } = params

		/**
		 * @type {Array<import('@tevm/block').Block>}
		 */
		const newBlocks = []
		/**
		 * @type {Map<import('@tevm/utils').Hex,Array<import('@tevm/receipt-manager').TxReceipt>>}
		 */
		const newReceipts = new Map()

		client.logger.debug({ blockCount }, 'processing txs')
		const pool = await client.getTxPool()
		const originalVm = await client.getVm()

		switch (client.status) {
			case 'MINING': {
				// wait for the previous mine to finish
				await new Promise((resolve) => {
					client.on('newBlock', async () => {
						if (client.status === 'MINING') {
							return
						}
						client.status = 'MINING'
						resolve(client)
					})
				})
				break
			}
			case 'INITIALIZING': {
				await client.ready()
				client.status = 'MINING'
				break
			}
			case 'SYNCING': {
				const err = new MisconfiguredClientError('Syncing not currently implemented')
				return maybeThrowOnFail(throwOnFail, { errors: [err] })
			}
			case 'STOPPED': {
				const err = new MisconfiguredClientError('Client is stopped')
				return maybeThrowOnFail(throwOnFail, { errors: [err] })
			}
			case 'READY': {
				client.status = 'MINING'
				break
			}
			default: {
				const err = new UnreachableCodeError(client.status)
				return maybeThrowOnFail(throwOnFail, { errors: [err] })
			}
		}

		const vm = await originalVm.deepCopy()
		const receiptsManager = await client.getReceiptsManager()

		for (let count = 0; count < blockCount; count++) {
			const parentBlock = await vm.blockchain.getCanonicalHeadBlock()

			let timestamp = Math.max(Math.floor(Date.now() / 1000), Number(parentBlock.header.timestamp))
			timestamp = count === 0 ? timestamp : timestamp + interval

			const blockBuilder = await vm.buildBlock({
				parentBlock,
				headerData: {
					timestamp,
					number: parentBlock.header.number + 1n,
					// The following 2 are currently not supported
					// difficulty: undefined,
					// coinbase,
					gasLimit: parentBlock.header.gasLimit,
					baseFeePerGas: parentBlock.header.calcNextBaseFee(),
				},
				blockOpts: {
					// Proof of authority not currently supported
					// cliqueSigner,
					// proof of work not currently supported
					//calcDifficultyFromHeader,
					//ck
					freeze: false,
					setHardfork: false,
					putBlockIntoBlockchain: false,
					common: vm.common,
				},
			})
			// TODO create a Log manager
			const orderedTx = await pool.txsByPriceAndNonce({ baseFee: parentBlock.header.calcNextBaseFee() })

			let index = 0
			// TODO we need to actually handle this
			const blockFull = false
			/**
			 * @type {Array<import('@tevm/receipt-manager').TxReceipt>}
			 */
			const receipts = []
			while (index < orderedTx.length && !blockFull) {
				const nextTx = /** @type {import('@tevm/tx').TypedTransaction}*/ (orderedTx[index])
				client.logger.debug(bytesToHex(nextTx.hash()), 'new tx added')
				const txResult = await blockBuilder.addTransaction(nextTx, {
					skipHardForkValidation: true,
				})
				if (txResult.execResult.exceptionError) {
					if (txResult.execResult.exceptionError.error === 'out of gas') {
						client.logger.debug(txResult.execResult.executionGasUsed, 'out of gas')
					}
					client.logger.debug(
						txResult.execResult.exceptionError,
						`There was an exception when building block for tx ${bytesToHex(nextTx.hash())}`,
					)
				}
				receipts.push(txResult.receipt)
				index++
			}
			await vm.stateManager.checkpoint()
			const createNewStateRoot = true
			await vm.stateManager.commit(createNewStateRoot)
			const block = await blockBuilder.build()
			await Promise.all([receiptsManager.saveReceipts(block, receipts), vm.blockchain.putBlock(block)])
			pool.removeNewBlockTxs([block])

			newBlocks.push(block)
			newReceipts.set(bytesToHex(block.hash()), receipts)

			const value = vm.stateManager._baseState.stateRoots.get(bytesToHex(block.header.stateRoot))

			if (!value) {
				throw new InternalError(
					'InternalError: State root not found in mineHandler. This indicates a potential inconsistency in state management.',
				)
			}

			originalVm.stateManager.saveStateRoot(block.header.stateRoot, value)
		}
		originalVm.blockchain = vm.blockchain
		originalVm.evm.blockchain = vm.evm.blockchain
		await originalVm.stateManager.setStateRoot(hexToBytes(vm.stateManager._baseState.getCurrentStateRoot()))

		client.status = 'READY'

		emitEvents(client, newBlocks, newReceipts)

		return { blockHashes: newBlocks.map((b) => bytesToHex(b.hash())) }
	}
