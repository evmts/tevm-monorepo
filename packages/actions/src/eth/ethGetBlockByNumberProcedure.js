import { createJsonRpcFetcher } from '@tevm/jsonrpc'
import { hexToBigInt } from '@tevm/utils'
import { blockToJsonRpcBlock } from '../utils/blockToJsonRpcBlock.js'

/**
 * Request handler for eth_getBlockByNumber JSON-RPC requests.
 * @param {import('@tevm/node').TevmNode} client
 * @returns {import('./EthProcedure.js').EthGetBlockByNumberJsonRpcProcedure}
 */
export const ethGetBlockByNumberJsonRpcProcedure = (client) => {
	return async (request) => {
		const vm = await client.getVm()
		const blockTagOrNumber = request.params[0]
		const block = await (() => {
			if (blockTagOrNumber.startsWith('0x')) {
				return vm.blockchain.getBlock(hexToBigInt(/** @type {import('@tevm/utils').Hex}*/ (blockTagOrNumber)))
			}
			return vm.blockchain.blocksByTag.get(/** @type {import('@tevm/utils').BlockTag}*/ (blockTagOrNumber))
		})()

		if (!block && client.forkTransport) {
			const fetcher = createJsonRpcFetcher(client.forkTransport)
			const res = await fetcher.request({
				jsonrpc: '2.0',
				id: request.id ?? 1,
				method: request.method,
				params: [blockTagOrNumber, request.params[1] ?? false],
			})
			if (res.error) {
				return {
					...(request.id ? { id: request.id } : {}),
					method: request.method,
					jsonrpc: request.jsonrpc,
					error: res.error,
				}
			}
			return {
				...(request.id ? { id: request.id } : {}),
				method: request.method,
				jsonrpc: request.jsonrpc,
				result: /** @type {any}*/ (res.result),
			}
		}
		if (!block) {
			return {
				...(request.id ? { id: request.id } : {}),
				method: request.method,
				jsonrpc: request.jsonrpc,
				error: {
					code: -32602,
					message: `Invalid block tag ${blockTagOrNumber}`,
				},
			}
		}
		const includeTransactions = request.params[1] ?? false
		const result = blockToJsonRpcBlock(block, includeTransactions)
		return {
			method: request.method,
			result,
			jsonrpc: '2.0',
			...(request.id ? { id: request.id } : {}),
		}
	}
}
