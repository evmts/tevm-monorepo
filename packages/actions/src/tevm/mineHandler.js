import { bytesToHex } from '@tevm/utils'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { validateMineParams } from '@tevm/zod'

// TODO Errors can leave us in bad states

/**
 * @param {import("@tevm/base-client").BaseClient} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').MineHandler}
 */
export const mineHandler =
  (client, options = {}) =>
    async ({ throwOnFail = options.throwOnFail ?? true, ...params } = {}) => {
      client.logger.debug({ throwOnFail, ...params }, 'mineHandler called with params')
      const errors = validateMineParams(params)
      if (errors.length > 0) {
        return maybeThrowOnFail(throwOnFail, { errors })
      }
      const { interval = 1, blockCount = 1 } = params

      /**
       * @type {Array<import('@tevm/utils').Hex>}
       */
      const blockHashes = []

      client.logger.debug({ blockCount }, 'processing txs')

      for (let count = 0; count < blockCount; count++) {
        const pool = await client.getTxPool()
        const vm = await client.getVm()
        const receiptsManager = await client.getReceiptsManager()
        const parentBlock = await vm.blockchain.getCanonicalHeadBlock()

        let timestamp = Math.max(Math.floor(Date.now() / 1000), Number(parentBlock.header.timestamp))
        timestamp = count === 0 ? timestamp : timestamp + interval

        const blockBuilder = await vm.buildBlock({
          parentBlock,
          headerData: {
            timestamp,
            number: parentBlock.header.number + 1n,
            // The following 2 are currently not supported
            // difficulty: undefined,
            // coinbase,
            gasLimit: parentBlock.header.gasLimit,
            baseFeePerGas: parentBlock.header.calcNextBaseFee(),
          },
          blockOpts: {
            // Proof of authority not currently supported
            // cliqueSigner,
            // proof of work not currently supported
            //calcDifficultyFromHeader,
            //ck
            freeze: false,
            setHardfork: false,
            putBlockIntoBlockchain: false,
            common: vm.common
          },
        })
        // TODO create a Log manager
        const orderedTx = await pool.txsByPriceAndNonce({ baseFee: parentBlock.header.calcNextBaseFee() })

        let index = 0
        let blockFull = false
        /**
         * @type {Array<import('@tevm/receipt-manager').TxReceipt>}
         */
        const receipts = []
        while (index < orderedTx.length && !blockFull) {
          const nextTx = /** @type {import('@tevm/tx').TypedTransaction}*/(orderedTx[index])
          client.logger.debug(bytesToHex(nextTx.hash()), 'new tx added')
          const txResult = await blockBuilder.addTransaction(nextTx, {
            skipHardForkValidation: true,
          })
          if (txResult.execResult.exceptionError) {
            if (txResult.execResult.exceptionError.error === 'out of gas') {
              client.logger.debug(txResult.execResult.executionGasUsed, 'out of gas')
            }
            client.logger.debug(txResult.execResult.exceptionError, `There was an exception when building block for tx ${bytesToHex(nextTx.hash())}`)
          }
          receipts.push(txResult.receipt)
          index++
        }
        await vm.stateManager.checkpoint()
        const createNewStateRoot = true
        await vm.stateManager.commit(createNewStateRoot)
        const block = await blockBuilder.build()
        await Promise.all([
          receiptsManager.saveReceipts(block, receipts),
          vm.blockchain.putBlock(block),
        ])
        pool.removeNewBlockTxs([block])

        blockHashes.push(bytesToHex(block.hash()))
      }
      return { blockHashes }
    }
