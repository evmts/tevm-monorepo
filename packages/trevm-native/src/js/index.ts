import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};