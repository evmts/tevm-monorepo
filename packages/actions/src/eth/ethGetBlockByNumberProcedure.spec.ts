import { type TevmNode, createTevmNode } from '@tevm/node'
import { beforeEach, describe, expect, it } from 'vitest'
import { mineHandler } from '../Mine/mineHandler.js'
import type { EthGetBlockByNumberJsonRpcRequest } from './EthJsonRpcRequest.js'
import { ethGetBlockByNumberJsonRpcProcedure } from './ethGetBlockByNumberProcedure.js'

let client: TevmNode

beforeEach(async () => {
	client = createTevmNode()
	await mineHandler(client)()
	await mineHandler(client)()
})

describe('ethGetBlockByNumberJsonRpcProcedure', () => {
	const normalizeBlockForSnapshot = (block: any) => {
		if (!block) return block
		return {
			...block,
			// Replace non-deterministic values with placeholders
			hash: typeof block.hash === 'string' && block.hash.startsWith('0x') ? '0x[HASH]' : block.hash,
			stateRoot:
				typeof block.stateRoot === 'string' && block.stateRoot.startsWith('0x') ? '0x[STATE_ROOT]' : block.stateRoot,
			timestamp:
				typeof block.timestamp === 'string' && block.timestamp.startsWith('0x') ? '0x[TIMESTAMP]' : block.timestamp,
		}
	}

	it('should return block details by number', async () => {
		const request: EthGetBlockByNumberJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_getBlockByNumber',
			id: 1,
			params: ['0x1', false], // Block number as hex
		}

		const response = await ethGetBlockByNumberJsonRpcProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_getBlockByNumber')
		expect(response.id).toBe(request.id as any)

		// Normalize the result before snapshot comparison
		const normalizedResult = normalizeBlockForSnapshot(response.result)
		expect(normalizedResult).toMatchSnapshot()
	})

	it('should include transactions if requested', async () => {
		const request: EthGetBlockByNumberJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_getBlockByNumber',
			id: 1,
			params: ['0x1', true], // Block number as hex
		}

		const response = await ethGetBlockByNumberJsonRpcProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_getBlockByNumber')
		expect(response.id).toBe(request.id as any)

		// Normalize the result before snapshot comparison
		const normalizedResult = normalizeBlockForSnapshot(response.result)
		expect(normalizedResult).toMatchSnapshot()
	})

	it('should handle requests without an id', async () => {
		const request: EthGetBlockByNumberJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_getBlockByNumber',
			params: ['0x1', false], // Block number as hex
		}

		const response = await ethGetBlockByNumberJsonRpcProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_getBlockByNumber')
		expect(response.id).toBeUndefined()

		// Normalize the result before snapshot comparison
		const normalizedResult = normalizeBlockForSnapshot(response.result)
		expect(normalizedResult).toMatchSnapshot()
	})

	it('should handle invalid block tag', async () => {
		const request: EthGetBlockByNumberJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_getBlockByNumber',
			id: 1,
			params: ['invalidTag' as any, false], // Invalid block tag
		}

		const response = await ethGetBlockByNumberJsonRpcProcedure(client)(request)
		expect(response.error).toBeDefined()
		expect(response.error).toMatchSnapshot()
	})

	it('should NOT return a Promise object in the result (reproduces bug)', async () => {
		const request: EthGetBlockByNumberJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_getBlockByNumber',
			id: 1,
			params: ['0x1', false], // Block number as hex
		}

		const response = await ethGetBlockByNumberJsonRpcProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()

		// This test reproduces the bug: the result should be a plain object, not a Promise
		expect(response.result).not.toBeInstanceOf(Promise)
		expect(typeof response.result).toBe('object')
		expect(response.result).toHaveProperty('number')
		expect(response.result).toHaveProperty('hash')
		expect(response.result).toHaveProperty('parentHash')
	})
})
