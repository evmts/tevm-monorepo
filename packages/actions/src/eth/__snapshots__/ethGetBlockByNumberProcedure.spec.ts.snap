// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ethGetBlockByNumberJsonRpcProcedure > should handle invalid block tag 1`] = `
{
  "code": -32602,
  "message": "Invalid block tag invalidTag",
}
`;

exports[`ethGetBlockByNumberJsonRpcProcedure > should handle requests without an id 1`] = `
{
  "baseFeePerGas": "0x7",
  "blobGasUsed": "0x0",
  "difficulty": "0x0",
  "excessBlobGas": "0x0",
  "extraData": "0x",
  "gasLimit": "0x1c9c380",
  "gasUsed": "0x0",
  "hash": "0x[HASH]",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "miner": "0x0000000000000000000000000000000000000000",
  "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "nonce": "0x0000000000000000",
  "number": "0x1",
  "parentBeaconBlockRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "parentHash": "0x1d230d3c25232b8bb73be506cb871ccb88ec60dfe77b5f58e1794293a0d62534",
  "receiptsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "requests": [],
  "requestsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "size": "0x65e",
  "stateRoot": "0x[STATE_ROOT]",
  "timestamp": "0x[TIMESTAMP]",
  "totalDifficulty": "0x0",
  "transactions": [],
  "transactionsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "uncles": [],
  "withdrawals": [],
  "withdrawalsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
}
`;

exports[`ethGetBlockByNumberJsonRpcProcedure > should include transactions if requested 1`] = `
{
  "baseFeePerGas": "0x7",
  "blobGasUsed": "0x0",
  "difficulty": "0x0",
  "excessBlobGas": "0x0",
  "extraData": "0x",
  "gasLimit": "0x1c9c380",
  "gasUsed": "0x0",
  "hash": "0x[HASH]",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "miner": "0x0000000000000000000000000000000000000000",
  "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "nonce": "0x0000000000000000",
  "number": "0x1",
  "parentBeaconBlockRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "parentHash": "0x1d230d3c25232b8bb73be506cb871ccb88ec60dfe77b5f58e1794293a0d62534",
  "receiptsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "requests": [],
  "requestsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "size": "0x65e",
  "stateRoot": "0x[STATE_ROOT]",
  "timestamp": "0x[TIMESTAMP]",
  "totalDifficulty": "0x0",
  "transactions": [],
  "transactionsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "uncles": [],
  "withdrawals": [],
  "withdrawalsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
}
`;

exports[`ethGetBlockByNumberJsonRpcProcedure > should return block details by number 1`] = `
{
  "baseFeePerGas": "0x7",
  "blobGasUsed": "0x0",
  "difficulty": "0x0",
  "excessBlobGas": "0x0",
  "extraData": "0x",
  "gasLimit": "0x1c9c380",
  "gasUsed": "0x0",
  "hash": "0x[HASH]",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "miner": "0x0000000000000000000000000000000000000000",
  "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "nonce": "0x0000000000000000",
  "number": "0x1",
  "parentBeaconBlockRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "parentHash": "0x1d230d3c25232b8bb73be506cb871ccb88ec60dfe77b5f58e1794293a0d62534",
  "receiptsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "requests": [],
  "requestsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "size": "0x65e",
  "stateRoot": "0x[STATE_ROOT]",
  "timestamp": "0x[TIMESTAMP]",
  "totalDifficulty": "0x0",
  "transactions": [],
  "transactionsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "uncles": [],
  "withdrawals": [],
  "withdrawalsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
}
`;
